EchoServer/EchoServer.cs

```diff
@@ -0,0 +1,107 @@
+using System.Net.Sockets;
+using System.Text;
+
+namespace EchoServer;
+
+public class EchoServer
+{
+    private Socket? _handler;
+    private NetworkStream? _stream;
+    public string? Data { get; private set; } = string.Empty;
+
+    public EchoServer()
+    {
+        Console.WriteLine("Initializing EchoServer.");
+    }
+
+    public void RunServer(Socket listener)
+    {
+        StartServer(listener);
+        Console.WriteLine("Stream created, beginning Echo...");
+        Echo();
+    }
+
+    public void StartServer(Socket listener)
+    {
+        try
+        {
+            Console.WriteLine("Accepting connections...");
+            _handler = listener.Accept();
+            Console.WriteLine("Connection Accepted! Opening stream...");
+            _stream = new NetworkStream(_handler);
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine(e.ToString());
+        }
+    }
+
+    
+    private void Echo()
+    {
+        string rawData = ReadStream();
+        Data = ProcessRawData(rawData);
+        string echoString = AssembleEchoString();
+        WriteStream(echoString);
+        CloseHandler();
+        
+    }
+
+    private string ReadStream()
+    {
+        string result = string.Empty;
+        do
+        {
+            byte[] buffer = new byte[16];
+            int bytesRead = _stream!.Read(buffer, 0, buffer.Length);
+            result += Encoding.UTF8.GetString(buffer, 0, bytesRead);
+        } while (_stream.DataAvailable);
+
+
+        return result;
+    }
+
+    private string ProcessRawData(string rawData)
+    {
+        string[] resultArray = rawData.Split("\r\n");
+        string result = resultArray.Last();
+
+        return result;
+    }
+
+    private string AssembleEchoString()
+    {
+        int contentLength = Data!.Length;
+        DateTime now = DateTime.Now;
+        string httpNow = now.ToUniversalTime().ToString("r"); 
+        
+        string result = @$"HTTP/1.1 200 OK
```

Comment:
Instead of hardcoding this, was there an option to dynamically generate those response headers since they're actually available in the Request headers coming in? Possibly in `ProcessRawData`? I know we were on a time crunch but is there the possibility of that?

Reply:
From the research I've done, no, there is no way to dynamically generate response headers without using an HTTP client that does all the socket stuff for me. If someone else is able to find a way to do this that doesn't mess with the whole point of the echo server, please let me know!

Reply:
I suppose you could read them from the request header and use the supplied http version. They might not match since the echo server is written to the HTTP/1.1 spec. This seems ok to me.